transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,154 @@
+
+_In cloud environments_, when hosting entities (e.g. containers, pods, virtual machines, etc.) are scaled down, their state is erased, e.g. their file system, ip address, and memory are deleted. This is exactly the situation where WildFly cannot guarantee that all transactions will be completed and a data integrity issue might occur.
+
+_1.2 There are additional complications when hosting entities crash (i.e. an unintentional shutdown occurs). This extreme scenario is beyond what WildFly can handle and the hosting cloud environment must provide the ability to recover the hosting entity from crashes._
+
+The state of the hosting entity (e.g. its file system, ip address, etc.) should be recovered by the cloud environment without the intervention of the user. In K8s, for example, the employment of StatefulSet creates a stable environment where crashes are handled automatically. If a particular cloud environment does not provide a guarantee similar to StatefulSet, it is not possible to employ WildFly’s Transaction Subsystem.
+
+*As a solution to solve problem 1.2 cannot be developed in WildFly (as it is impossible to control the lifecycle of the host from WildFly), this proposal is only concerned with problem 1.1.*
+
+== Issue Metadata
+
+=== Issue
+
+* [WFLY-17742](https://issues.jboss.org/browse/WFLY-17742)
+
+=== Dev Contacts
+
+* mailto:{jfinelli@redhat.com}[{Manuel Finelli}]
+
+=== QE Contacts
+
+* mailto:{istraka@redhat.com}[{Ivan Straka}]
+
+=== Testing By
+
+* [ ] Engineering
+* [X] QE
+
+=== Affected Projects or Components
+
+* [wildfly-core](https://github.com/wildfly/wildfly-core)
+* [wildfly](https://github.com/wildfly/wildfly)
+* [narayana](https://github.com/jbosstm/narayana)
+
+=== Relevant Installation Types
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* The Transactions subsystem should become aware of how many in-doubt transactions there are in the Object Store and delay its suspension as long as all transactions are completed. In case (some of the) in-doubt transactions fail, the Transactions subsystem should wait as long as those transactions are resolved. This would guarantee data integrity in all scenarios, including situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
```

Comment:
Have you considered heuristic transactions? IIUC admin need to step in and handle it by hand. Would it make sense if the suspend operation also provided an option (a prameter) to fail if there are heuristics transaction?

Reply:
I do definitely agree with the need to consider heuristic transactions - thank you @istraka.

I am not sure whether actually changing the design is appropriate and in particular if that were to be decided as a course of action I am not certain enough of the implications of failing graceful shutdown to be able to comment if failing graceful shutdown would be appropriate. My guess is that alone would not be "correct", presumably the logs should also be somehow made available, and the node name should remain locked until the transactions are resolved (and if that is the case having graceful shutdown fail seems it might not help from an efficiency point of view) but acknowledging the topic is appropriate in this design document and possibly addressing it (here or as an improvement for somewhere else).

Reply:
AFAIU, heuristic transactions are still in-doubt transactions, that's why I didn't explicitly include heuristic transactions in the requirements. Do you think I should make heuristic transactions more visible in the requirements? @istraka @tomjenkinson 

Just as a side note...in the design presented here, the original logic was:
1) Heuristic transactions will block WildFly's graceful shutdown like any other in-doubt transaction
2) WildFly's notification system will be employed to notify the user that heuristic transactions are blocking WildFly's graceful shutdown

Reply:
They are different to in-doubt transactions as they are somehow-complete transactions but where the resource manager made an outcome decision rather than under the direction of the transaction manager. Their log records are not removed as they are heuristically completed and this is right for us to do that. Resource managers must retain information about the heuristically completed branches. A good reference entry into the topic can be found in XA specification section 2.3.4.

BTW I am not saying it necessarily needs to change the design still, mainly trying to help clarify in regards your response `heuristic transactions are still in-doubt transactions`. In-doubt are different to me as the decision should still be under the control of the transaction manager.

Calling it out as a distinct notification type might be appropriate. The action for dealing with a heuristic transaction will be different to in-doubt transactions as it will require manual intervention. @mmusgrov could provide some more input on this, we could already have some documented guidance showinf how users could approach dealing with a heuristically completed transaction branch (and therefore transaction).

Reply:
Yes, in doubt means that we have called prepare and all participants have prepared and we have written a log record, the doubt is about the final outcome: we may call commit but we are not sure that they will all commit successfully.

But a heuristic occurs when a resource commits or rolls back autonomously (ie not under the instruction of the TM) - this applies even if the resource does the same thing that the TM will eventually ask it to do (for instance if it commits and then the TM tells it to commit then it is still a heuristic since the commit may have prematurely released locks).

The EAP documentation does have a section about manually resolving heuristics but in general they are unrecoverable.

Reply:
Thank you very much for clarifying the difference between in-doubt and heuristic transactions +1. I'll add a point about heuristic transactions to the hard requirements then